src/Microsoft.OpenApi/Services/OpenApiSecurityRequirementComparer.cs

```diff
@@ -0,0 +1,96 @@
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Update,
+                        SourceValue = sourceSecurityRequirement,
+                        TargetValue = targetSecurityRequirement,
+                        OpenApiComparedElementType = typeof(OpenApiSecurityRequirement),
+                        Pointer = comparisonContext.PathString
+                    });
+
+                return;
+            }
+
+            var newSecuritySchemesInTarget = targetSecurityRequirement.Keys
+                .Where(targetReq => sourceSecurityRequirement.Keys.All(
+                    sourceReq => sourceReq.Reference.Id != targetReq.Reference.Id)).ToList();
+
+            foreach (var newSecuritySchemeInTarget in newSecuritySchemesInTarget)
+            {
+                WalkAndAddOpenApiDifference(
+                    comparisonContext,
+                    newSecuritySchemeInTarget.Reference.Id,
+                    new OpenApiDifference
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Add,
+                        TargetValue = new KeyValuePair<OpenApiSecurityScheme, IList<string>>(
+                            newSecuritySchemeInTarget,
+                            targetSecurityRequirement[newSecuritySchemeInTarget]),
+                        OpenApiComparedElementType = typeof(KeyValuePair<OpenApiSecurityScheme, IList<string>>)
+                    });
+            }
+
+            foreach (var sourceSecurityScheme in sourceSecurityRequirement.Keys)
+            {
+                var targetSecurityScheme =
+                    targetSecurityRequirement.Keys.FirstOrDefault(
+                        i => i.Reference.Id == sourceSecurityScheme.Reference.Id);
+
+                if (targetSecurityScheme == null)
+                {
+                    WalkAndAddOpenApiDifference(
+                        comparisonContext,
+                        sourceSecurityScheme.Reference.Id,
+                        new OpenApiDifference
+                        {
+                            OpenApiDifferenceOperation = OpenApiDifferenceOperation.Remove,
+                            SourceValue = new KeyValuePair<OpenApiSecurityScheme, IList<string>>(
+                                sourceSecurityScheme,
+                                sourceSecurityRequirement[sourceSecurityScheme]),
+                            OpenApiComparedElementType = typeof(KeyValuePair<OpenApiSecurityScheme, IList<string>>)
+                        });
+                }
+                else
+                {
+                    WalkAndCompare(comparisonContext,
+                        sourceSecurityScheme.Reference.Id,
+                        () => comparisonContext
+                            .GetComparer<OpenApiSecurityScheme>()
+                            .Compare(sourceSecurityScheme, targetSecurityScheme, comparisonContext));
```

Comment:
I don't think we should be comparing the OpenApiSecurityScheme. Shouldn't we be comparing the list of strings in the value side?

Reply:
why not compare Security scheme? if source and target reference two different types of security schemes there should be a diff for it.

---
In reply to: [223516846](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223516846) [](ancestors = 223516846)

Reply:
That said i was thinking whether to compare scope list or not, and if i do compare how the pointer should be ?

e.g. if source is 


---
In reply to: [223852204](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223852204) [](ancestors = 223852204,223516846)

Reply:
- We should not compare SecurityScheme here because we only use its name as a key. From the OpenAPI document standpoint, only the key (id of the ref) will show up. Even if we manually craft the DOM to actually be different (say, add some random other things into the Security Scheme), the resulting serialized document will be exactly the same for this SecurityRequirement object.

- Ideally, diff in the above case should be Remove /scheme1/0. However, from the way you handle the Ordered List (in OpenApiOrderedListComparer), this can be
  
   - Update /scheme1/0 from scope1 to scope2
   - Update /scheme1/1 from scope2 to scope3
   - Remove /scheme1/2

  which is also okay (and correct).



---
In reply to: [223852979](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223852979) [](ancestors = 223852979,223852204,223516846)

Reply:
If we want to be really fancy, we can use an algo for "minimum edit distance" to find the best list of update/add/remove (think about one scope being one letter in the original problem).

---
In reply to: [223900360](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223900360) [](ancestors = 223900360,223852979,223852204,223516846)

Reply:
I will have a separate PR to compare scopes

---
In reply to: [224182344](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r224182344) [](ancestors = 224182344,223900360,223852979,223852204,223516846)